crates/bevy_ecs/hecs/src/dynamic_query.rs

```diff
@@ -0,0 +1,527 @@
+            let ptr = archetype.get_dynamic(component.id, component.layout.size(), 0);
+
+            if let Some(ptr) = ptr {
+                matches_any = true;
+
+                fetch
+                    .immutable
+                    .push(NonNull::new_unchecked(ptr.as_ptr().add(offset)));
+            } else {
+                missing_any = true;
+            }
+        }
+
+        // Query mutable components
+        for component in &self.mutable {
+            let ptr = archetype.get_dynamic(component.id, component.layout.size(), 0);
+
+            if let Some(ptr) = ptr {
+                matches_any = true;
+
+                fetch
+                    .mutable
+                    .push(NonNull::new_unchecked(ptr.as_ptr().add(offset)));
+            } else {
+                missing_any = true;
+            }
+        }
+
+        if matches_any && !missing_any {
+            Some(fetch)
+        } else {
+            None
+        }
+    }
+}
+
+pub struct DynamicQueryFetch<'query> {
+    query: &'query DynamicQuery,
+    entity: Option<NonNull<Entity>>,
+    immutable: Vec<NonNull<u8>>,
+    mutable: Vec<NonNull<u8>>,
+}
+
+impl<'query> DynamicQueryFetch<'query> {
+    fn new(query: &'query DynamicQuery) -> Self {
+        Self {
+            query,
+            entity: Default::default(),
+            immutable: Default::default(),
+            mutable: Default::default(),
+        }
+    }
+
+    /// # Safety
+    ///
+    /// Index must be within bounds
+    pub(crate) unsafe fn fetch<'a>(&mut self, index: usize) -> DynamicQueryResult<'a> {
+        let mut query_result = DynamicQueryResult {
+            // FIXME: Does this entity need to be cloned to be safe, or is it safe just to
+            // dereference it?
+            entity: self.entity.map(|x| *x.as_ptr().add(index)),
+            immutable: Vec::with_capacity(self.immutable.len()),
+            mutable: Vec::with_capacity(self.mutable.len()),
+        };
+
+        for (component, ptr) in self.query.immutable.iter().zip(&self.immutable) {
+            query_result.immutable.push(&*slice_from_raw_parts(
+                ptr.as_ptr().add(index * component.layout.size()),
+                component.layout.size(),
+            ));
+        }
+
+        for (component, ptr) in self.query.mutable.iter().zip(&self.mutable) {
+            query_result.mutable.push(&mut *slice_from_raw_parts_mut(
+                ptr.as_ptr().add(index * component.layout.size()),
+                component.layout.size(),
+            ));
+        }
+
+        query_result
+    }
+}
+
+/// The data returned by [`DynamicFetch::fetch`]
+#[derive(Default)]
+pub struct DynamicQueryResult<'a> {
+    /// The entity, if requested in the Fetch
+    pub entity: Option<Entity>,
+    /// The immutable data for the immutable components requested
+    pub immutable: Vec<&'a [u8]>,
+    /// The mutable data for the immutable components requested
+    pub mutable: Vec<&'a mut [u8]>,
```

Comment:
Could we use a slice-of-slices here rather than owned storage? The `Borrow` structure might be a good place to stash reusable memory for that.

Reply:
Yeah, I'll try that out, I was debating how to handle that efficiently. Earlier I had used fixed-width arrays, but I wasn't sure if that was a good idea earlier.

Reply:
If I understand correctly, if we stuck the reusable memory in `DynamicBorrow`, it wouldn't be possible to later to chunked iteration, because all of the different `DynamicQueryFetch` instances that are being used in different threads would be writing to the same memory. I may not be understanding that correctly, though, and maybe we just cross that bridge when we get there.

Any thoughts on using an iterator instead of a slice? So it would be like `pub immutable: impl Iterator<Item = &'a [u8]>`.

Now I'm not sure if for convenience most people would just end up doing a `collect()` on that iterator which would cause an allocation anyway and we'd be in the same boat as before where we might as well have stuck it in a `Vec`.

Just an idea.

Reply:
OK, I found out that my understanding of the lifetimes was wrong and that an iterator wouldn't work because it would have to borrow from `QueryFetch` and that the only place that the storage *could* be borrowed from is either the `Borrow` or the `Query`.

---

Anyway, I put some effort into trying to use storage in `DynamicBorrow` like you suggested, and the issue is that things get hairy when you need a mutable reference to `DynamicBorrow` that goes all the way down to the `DynamicFetch`. My attempt starts to break down in the `DynamicQueryIter` ( at least ) because it requires mutating `DynamicBorrow` now to be able to access its storage, and I run into iterator problems because you apparently [can't iterate mutably without `unsafe`](https://users.rust-lang.org/t/iterator-lifetime-error-only-when-using-a-mutable-reference/50460/6?u=zicklag).

It also complicates the code quite a bit. This PR is actually the first time I've really gotten into the meat of Rust lifetimes, so maybe I'm missing a relatively simple way to do it, but I'm feeling like the juice isn't worth the squeeze here. Maybe this is something that can be improved on later? I'll add a TODO to the code.